src/toxcore/dht_node.rs

```diff
@@ -287,141 +343,292 @@ mod test {
         quickcheck(with_nodes as fn(Vec<PackedNode>));
     }
 
+    // DhtNode::create_ping_req()
+
+    #[test]
+    fn dht_node_create_ping_req_test() {
+        let alice = DhtNode::new().unwrap();
+        let (bob_pk, bob_sk) = gen_keypair();
+        let (_, eve_sk) = gen_keypair();
+        let packet1 = alice.create_ping_req(&bob_pk);
+        assert_eq!(*alice.dht_public_key, packet1.sender_pk);
+        assert_eq!(PacketKind::PingReq, packet1.kind());
+
+        let packet2 = alice.create_ping_req(&bob_pk);
+        assert_ne!(packet1, packet2);
+
+        // eve can't decrypt it
+        assert_eq!(None, packet1.get_packet::<PingReq>(&eve_sk));
+
+        let payload1: PingReq = packet1.get_packet(&bob_sk)
+            .expect("failed to get payload1");
+        let payload2: PingReq = packet2.get_packet(&bob_sk)
+            .expect("failed to get payload2");
+        assert_ne!(payload1.id(), payload2.id());
+    }
+
+
+    // DhtNode::request_ping()
+
+    #[test]
+    fn dht_node_request_ping_test() {
+        // bob creates & sends PingReq to alice
+        // received PingReq has to be succesfully decrypted
+        node_socket!(core, handle,
+            alice, alice_socket,
+            bob, bob_socket);
+
+        let alice_pn = PackedNode::new(true,
+            alice_socket.local_addr().unwrap(),
+            &alice.dht_public_key);
+
+        let mut recv_buf = [0; MAX_UDP_PACKET_SIZE];
+
+        let bob_framed = bob_socket.framed(ToxCodec);
+        let bob_request = bob.request_ping(bob_framed, &alice_pn);
+
+        let future_recv = alice_socket.recv_dgram(&mut recv_buf[..]);
+        let future_recv = add_timeout!(future_recv, &handle);
+        handle.spawn(bob_request.then(|_| ok(())));
+
+        let received = core.run(future_recv).unwrap();
+        let (_alice_socket, recv_buf, size, _saddr) = received;
+        assert!(size != 0);
+
+        let recv_packet = DhtPacket::from_bytes(&recv_buf[..size]).unwrap();
+        let payload: PingReq = recv_packet
+            .get_packet(&alice.dht_secret_key)
+            .expect("Failed to decrypt payload");
```

Comment:
+ add case for Eve

Reply:
This is checked in test above, which tests function deals with crypto, should the check still be added here?

Reply:
no. you may add get_packet<DIFFERENT_TYPE> to check that the payload cannot be parsed.

Reply:
Done for the packets that can't be parsed as other packets.